File: patterns/04-concurrency-patterns/01-bounded-producer-consumer/js/bounded_producer_consumer.mjs
export class PatternError extends Error { constructor(code) { super(code); this.code = code; } }

export function evaluate(input) {
    if (input.capacity < 0) throw new PatternError("invalid_capacity");
    const queue=[], consumed=[], receipts=[]; let closed=false;
    for (const action of input.actions ?? []) {
        if (action.op === "produce") {
            if (closed) throw new PatternError("produce_after_close");
            if (queue.length >= input.capacity) receipts.push(`backpressured:${action.item}`); else { queue.push(action.item); receipts.push(`queued:${action.item}`); }
        } else if (action.op === "consume") {
            if (queue.length) { const item=queue.shift(); consumed.push(item); receipts.push(`consumed:${item}`); } else receipts.push(closed ? "end" : "empty");
        } else if (action.op === "close") {
            if (closed) throw new PatternError("already_closed"); closed=true; receipts.push("closed");
        } else throw new PatternError("unknown_action");
    }
    return {receipts,consumed,remaining:queue,closed};
}

File: patterns/04-concurrency-patterns/01-bounded-producer-consumer/go/bounded_producer_consumer.go
package boundedproducerconsumer

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

func Evaluate(input map[string]any) (any, error) {
	capacity := int(input["capacity"].(float64))
	if capacity < 0 {
		return nil, &PatternError{code: "invalid_capacity"}
	}
	queue := []string{}
	consumed := []string{}
	receipts := []string{}
	closed := false
	actions, _ := input["actions"].([]any)
	for _, item := range actions {
		action := item.(map[string]any)
		operation := action["op"].(string)
		switch operation {
		case "produce":
			value := action["item"].(string)
			if closed {
				return nil, &PatternError{code: "produce_after_close"}
			}
			if len(queue) >= capacity {
				receipts = append(receipts, "backpressured:"+value)
			} else {
				queue = append(queue, value)
				receipts = append(receipts, "queued:"+value)
			}
		case "consume":
			if len(queue) > 0 {
				value := queue[0]
				queue = queue[1:]
				consumed = append(consumed, value)
				receipts = append(receipts, "consumed:"+value)
			} else if closed {
				receipts = append(receipts, "end")
			} else {
				receipts = append(receipts, "empty")
			}
		case "close":
			if closed {
				return nil, &PatternError{code: "already_closed"}
			}
			closed = true
			receipts = append(receipts, "closed")
		default:
			return nil, &PatternError{code: "unknown_action"}
		}
	}
	return map[string]any{"receipts": receipts, "consumed": consumed, "remaining": queue, "closed": closed}, nil
}
